### PR TITLE
create, read and update private urls

### LINKED ISSUE #10: add a checkbox to turn links private
set `private: true` in model

private links should not redirect unless user logs in (private links are to be used on workshops where we need info of user before we show them info)

### PR DESCRIPTION
**Pull request by jatinkatyal13**
Fixes #10

### CODE REVIEW

File: src/controllers/events.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { Events, EventAttributes, URLAttributes, UserAttributes } from '../db'
+
+export const createEvent = async (url: URLAttributes, req, user: UserAttributes | undefined) => {
+    const event = Events.create({
+        code: url.code,
+        fromIP: req.headers['x-forwarded-for'] || req.connection.remoteAddress,
+        fromURL: req.query.refer || null,
```
Comment (championswimmer) [inline src/controllers/events.ts]: referer

File: src/controllers/events.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+        userId: user ? user.id : undefined
+    })
+    if (!event) {
+        throw new Error('Could not create event')
```
Comment (championswimmer) [inline src/controllers/events.ts]: do not throw, just capture in Raven
this should not make the app crash, this is not critical functionality

Comment (championswimmer) [reply src/controllers/events.ts]: or catch it later ( i have commented below where you can catch)

File: src/routes/pages/expand.ts [hunk 3]
```diff
@@ -2,23 +2,30 @@ import { Router } from 'express'
...
+    if (url.private) {
+      if (!req.isAuthenticated()) {
+        return res.render('pages/urls/private')
+      }
+    }
+
     res.redirect(url.longUrl)
     // Redirect first, then handle hit increment later
-    // TODO: Generate event too
+    createEvent(url, req, req.user)
```
Comment (championswimmer) [inline src/routes/pages/expand.ts]: should catch and capture into Raven just like the increment step

File: views/pages/urls/new.hbs [hunk 4]
```diff
@@ -39,6 +39,18 @@
             Shortcode has to be less than 9 characters
           </small>
         </div>
+        {{#ifcontains "admin,employee" user.role}}
```
Comment (championswimmer) [inline views/pages/urls/new.hbs]: i've updated your role in accounts to employee so you can try out these type of things easily.

Comment (jatinkatyal13) [reply views/pages/urls/new.hbs]: Until now I just updated the role directly in the db 😜

File: src/controllers/events.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+export const createEvent = async (
+  url: URLAttributes,
+  req,
+  user: UserAttributes | undefined,
+) => {
+  try {
+    const event = Events.create({
+      code: url.code,
+      fromIP: req.headers['x-forwarded-for'] || req.connection.remoteAddress,
```
Comment (jatinkatyal13) [inline src/controllers/events.ts]: Any better way of doing this? @championswimmer

Comment (championswimmer) [reply src/controllers/events.ts]: https://www.npmjs.com/package/request-ip this handles all cases

**Comment by championswimmer** [comment]
also, run `yarn run lint:fix` once before committing. It reformats all documents as per the linter

File: views/pages/urls/url.hbs [hunk 6]
```diff
@@ -34,4 +56,33 @@
     </div>
   </div>
   <div class="col"></div>
-</div>
\ No newline at end of file
+</div>
+
+<script>
+  // Show an element
+  function show (elem) {
```
Comment (championswimmer) [inline views/pages/urls/url.hbs]: since we are already using jQuery, just use jQuery's `hide()` and `show()` handlers

**Pull request merged**